File: python/y2020/day01.py
from aoc.core import solve, AocInput
from itertools import product
from functools import reduce


def part1(input: AocInput) -> int:
    numbers = list(map(int, input.lines))
    entries = product(numbers, numbers)
    return reduce(lambda a, b: a * b, next(filter(lambda e: sum(e) == 2020, entries)))


def part2(input: AocInput) -> int:
    numbers = list(map(int, input.lines))
    entries = product(numbers, numbers, numbers)
    return reduce(lambda a, b: a * b, next(filter(lambda e: sum(e) == 2020, entries)))


def main() -> None:
    YEAR = "y2020"
    DAY = "Day01"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1)
    solve(YEAR, DAY, part2)

File: dotnet/2020/AdventOfCode.Y2020/Day01.cs
using AdventOfCode.Lib;

namespace AdventOfCode.Y2020;

public sealed class Day01 : IAocDay<int>
{
    public static int Part1(AocInput input) =>
        input.Lines.Select(int.Parse).ToArray()
        | (numbers => (
                from i in numbers
                from j in numbers
                select new[] { i, j }
            )
            .First(n => n.Sum() is 2020)
            .Aggregate((i, j) => i * j)
        );

    public static int Part2(AocInput input) =>
        input.Lines.Select(int.Parse).ToArray()
        | (numbers => (
                from i in numbers
                from j in numbers
                from k in numbers
                select new[] { i, j, k }
            )
            .First(n => n.Sum() is 2020)
            .Aggregate((i, j) => i * j)
        );
}